pkg/storegateway/indexheader/index/postings.go

```diff
@@ -0,0 +1,514 @@
+		}
+		// In any case lastValOffset is unknown as don't have next posting anymore. Guess from TOC table.
+		// In worst case we will overfetch a few bytes.
+		t.postings[lastKey[0]].lastValOffset = int64(indexLastPostingEnd) - crc32.Size
+	}
+	// Trim any extra space in the slices.
+	for k, v := range t.postings {
+		l := make([]postingOffset, len(v.offsets))
+		copy(l, v.offsets)
+		t.postings[k].offsets = l
+	}
+
+	return &t, nil
+}
+
+// readOffsetTable reads an offset table and at the given position calls f for each
+// found entry. If f returns an error it stops decoding and returns the received error.
+func readOffsetTable(factory *stream_encoding.DecbufFactory, tableOffset int, f func([]string, uint64, int) error) error {
+	d := factory.NewDecbufAtChecked(tableOffset, castagnoliTable)
+	defer factory.Close(d)
+
+	startLen := d.Len()
+	cnt := d.Be32()
+
+	for d.Err() == nil && d.Len() > 0 && cnt > 0 {
+		offsetPos := startLen - d.Len()
+		keyCount := d.Uvarint()
+		// The Postings offset table takes only 2 keys per entry (name and value of label),
+		// and the LabelIndices offset table takes only 1 key per entry (a label name).
+		// Hence setting the size to max of both, i.e. 2.
+		keys := make([]string, 0, 2)
+
+		for i := 0; i < keyCount; i++ {
+			keys = append(keys, d.UvarintStr())
+		}
+		o := d.Uvarint64()
+		if d.Err() != nil {
+			break
+		}
+		if err := f(keys, o, offsetPos); err != nil {
+			return err
+		}
+		cnt--
+	}
+	return d.Err()
+}
+
+func (t *PostingOffsetTableV1) PostingsOffset(name string, values ...string) ([]index.Range, error) {
+	rngs := make([]index.Range, 0, len(values))
+	e, ok := t.postings[name]
+	if !ok {
+		return nil, nil
+	}
+	for _, v := range values {
+		rng, ok := e[v]
+		if !ok {
+			continue
+		}
+		rngs = append(rngs, rng)
+	}
+	return rngs, nil
+}
+
+func (t *PostingOffsetTableV1) LabelValues(name string, filter func(string) bool) ([]string, error) {
+	e, ok := t.postings[name]
+	if !ok {
+		return nil, nil
+	}
+	values := make([]string, 0, len(e))
+	for k := range e {
+		if filter == nil || filter(k) {
+			values = append(values, k)
+		}
+	}
+	sort.Strings(values)
+	return values, nil
+}
+
+func (t *PostingOffsetTableV1) LabelNames() ([]string, error) {
+	labelNames := make([]string, 0, t.LabelNameCount())
+
+	err := t.ForEachLabelName(func(name string) error {
+		labelNames = append(labelNames, name)
+		return nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	sort.Strings(labelNames)
+
+	return labelNames, nil
+}
+
+func (t *PostingOffsetTableV1) ForEachLabelName(f func(name string) error) error {
+	allPostingsKeyName, _ := index.AllPostingsKey()
+
+	for name := range t.postings {
+		if name == allPostingsKeyName {
+			continue
+		}
+
+		if err := f(name); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (t *PostingOffsetTableV1) LabelNameCount() int {
```

Comment:
Two nits:
1. No need to export it
2. The comment below should be a function comment

_Same comment applies to `PostingOffsetTableV2`._